src/include/storage/index/bplustree.h

```diff
@@ -0,0 +1,2554 @@
+/**
+ *  Base class for BPlusTree that stores common data, inherited by the BPlusTree class. This
+ *  class stores the threshold parameters specific to the B+ Tree.
+ */
+class BPlusTreeBase {
+ public:
+  /**
+   * @return inner_node_size_upper_threshold
+   */
+  int GetInnerNodeSizeUpperThreshold() const { return inner_node_size_upper_threshold_; }
+  /**
+   * @return inner_node_size_lower_threshold
+   */
+  int GetInnerNodeSizeLowerThreshold() const { return inner_node_size_lower_threshold_; }
+  /**
+   * @return leaf_node_size_upper_threshold
+   */
+  int GetLeafNodeSizeUpperThreshold() const { return leaf_node_size_upper_threshold_; }
+  /**
+   * @return leaf_node_size_lower_threshold
+   */
+  int GetLeafNodeSizeLowerThreshold() const { return leaf_node_size_lower_threshold_; }
+
+ public:
+  /**
+   * @param inner_node_size_upper_threshold upper size threshold for inner node split to be assigned to this tree
+   */
+  void SetInnerNodeSizeUpperThreshold(int inner_node_size_upper_threshold) {
+    inner_node_size_upper_threshold_ = inner_node_size_upper_threshold;
+  }
+
+  /**
+   * @param inner_node_size_lower_threshold lower size threshold for inner node removal to be assigned to this tree
+   */
+  void SetInnerNodeSizeLowerThreshold(int inner_node_size_lower_threshold) {
+    inner_node_size_lower_threshold_ = inner_node_size_lower_threshold;
+  }
+
+  /**
+   * @param leaf_node_size_upper_threshold upper size threshold for leaf node split to be assigned to this tree
+   */
+  void SetLeafNodeSizeUpperThreshold(int leaf_node_size_upper_threshold) {
+    leaf_node_size_upper_threshold_ = leaf_node_size_upper_threshold;
+  }
+
+  /**
+   * @param leaf_node_size_lower_threshold lower size threshold for leaf node removal to be assigned to this tree
+   */
+  void SetLeafNodeSizeLowerThreshold(int leaf_node_size_lower_threshold) {
+    leaf_node_size_lower_threshold_ = leaf_node_size_lower_threshold;
+  }
+
+ protected:
+  /** upper size threshold for inner node split */
+  int inner_node_size_upper_threshold_ = INNER_NODE_SIZE_UPPER_THRESHOLD;
+  /** lower size threshold for inner node removal */
+  int inner_node_size_lower_threshold_ = INNER_NODE_SIZE_LOWER_THRESHOLD;
+
+  /** upper size threshold for leaf node split */
+  int leaf_node_size_upper_threshold_ = LEAF_NODE_SIZE_UPPER_THRESHOLD;
+  /** lower size threshold for leaf node removal */
+  int leaf_node_size_lower_threshold_ = LEAF_NODE_SIZE_LOWER_THRESHOLD;
+
+ public:
+  /**
+   * Constructor
+   */
+  BPlusTreeBase() = default;
+
+  /**
+   * Destructor
+   */
+  ~BPlusTreeBase() = default;
+};
+
+/**
+ * Implementation of a B+ Tree index using latch crabbing.
+ *
+ * Concurrency:
+ *  Each node (leaf and inner) contains one shared latch to gain access to the node.
+ *  Writers are prioritized in this lock. Additionally, to guarantee safety of the root pointer of the
+ *  tree there is an additional latch for the root (root has 2 latches now). Latches are acquired from
```

Comment:
Can you write something more specifically about why the root pointer needs two latches for safety? It's not clear to me which case only having one latch for the root is unsafe.

Reply:
It's in the case where the root is being replaced, right? If so, it'd be nice to make that clearer in the comment.

Reply:
Done